rem/DB.swift

```diff
@@ -215,76 +225,98 @@ class DatabaseManager {
-            .limit(limit)
-            .select(allText[frameId], frames[activeApplicationName], frames[timestamp])
+            .limit(limit, offset: offset)
 
-        var results: [(Int64, String?, String?, Date)] = []
+        var results: [(Int64, String?, String?, Date, String, Int64)] = []
         do {
             for row in try db.prepare(query) {
-                let frameId = row[allText[frameId]]
+                let frameId = row[frames[id]]
                 let applicationName = row[frames[activeApplicationName]]
                 let timestamp = row[frames[timestamp]]
-                results.append((frameId, nil, applicationName, timestamp))
+                let filePath = row[videoChunks[filePath]]
+                let offsetIndex = row[frames[offsetIndex]]
+                results.append((frameId, nil, applicationName, timestamp, filePath, offsetIndex))
             }
         } catch {
             print("Error fetching recent results: \(error)")
         }
         return results
     }
     
-    func getImage(index: Int64) -> CGImage? {
+    func getImage(index: Int64, maxSize: CGSize? = nil) -> CGImage? {
         guard let frameData = DatabaseManager.shared.getFrame(forIndex: index) else { return nil }
         
         let videoURL = URL(fileURLWithPath: frameData.filePath)
-        if !FileManager.default.fileExists(atPath: videoURL.path) {
-            return nil
-        }
-        
-        return extractFrame(from: videoURL, frameOffset: frameData.offsetIndex)
+        return extractFrame(from: videoURL, frameOffset: frameData.offsetIndex, maxSize: maxSize)
     }
     
-    func extractFrame(from videoURL: URL, frameOffset: Int64) -> CGImage? {
+    func getImages(filePath: String, frameOffsets: [Int64], maxSize: CGSize? = nil) -> AVAssetImageGenerator.Images {
+        let videoURL = URL(fileURLWithPath: filePath)
+        return extractFrames(from: videoURL, frameOffsets: frameOffsets, maxSize: maxSize)
+    }
+    
+    func extractFrame(from videoURL: URL, frameOffset: Int64, maxSize: CGSize? = nil) -> CGImage? {
         let asset = AVURLAsset(url: videoURL)
         let generator = AVAssetImageGenerator(asset: asset)
+        if let maxSize = maxSize {
+            generator.maximumSize = maxSize
+        }
         generator.appliesPreferredTrackTransform = true
         generator.requestedTimeToleranceBefore = CMTime.zero;
         generator.requestedTimeToleranceAfter = CMTime.zero;
         
         do {
-            let a = CMTime(value: frameOffset, timescale: 25)
+            let a = CMTime(value: frameOffset, timescale: DatabaseManager.FPS)
             let aI = try generator.copyCGImage(at: a, actualTime: nil)
             return aI
         } catch {
             print("Error extracting frame \(videoURL): \(error)")
             return nil
         }
     }
+    
+    func extractFrames(from videoURL: URL, frameOffsets: [Int64], maxSize: CGSize? = nil) -> AVAssetImageGenerator.Images {
```

Comment:
batch retrieval. so really we should be doing this in timeline too, but #7 should be much more easily solvable with this. "are we near a new chunk? load the whole thing"